sodym/mfa_system.py

```diff
@@ -0,0 +1,172 @@
+from abc import ABC, abstractmethod
+import numpy as np
+from typing import Dict
+from .mfa_definition import MFADefinition
+from .named_dim_arrays import Flow, Process, Parameter, NamedDimArray
+from .stocks import Stock
+from .data_reader import DataReader
+
+
+class MFASystem(ABC):
+    """An MFASystem class handles the definition, setup and calculation of a Material Flow Analysis system, which
+    consists of a set of processes, flows, stocks defined over a set of dimensions. 
+    For the concrete definition of the system, a subclass of MFASystem must be implemented.
+
+    **Example**
+
+    Define your MFA System:
+
+    >>> from sodym import MFASystem
+    >>> class CustomMFA(MFASystem):
+    >>>     def set_up_definition(self) -> MFADefinition:
+    >>>         # define the model here
+    >>>     def compute(self):
+    >>>         # do some computations on the CustomMFA attributes: stocks and flows
+
+    Initialize and run your MFA System model:
+
+    >>> from sodym import ExampleDataReader
+    >>> data_reader = ExampleDataReader(...)
+    >>> mfa = MFASystem(data_reader=data_reader)
+    >>> mfa.compute()
+
+    MFA flows, stocks and parameters are defined as instances of subclasses of :py:class:`sodym.named_dim_arrays.NamedDimArray`.
+    Dimensions are managed with the :py:class:`sodym.dimensions.Dimension` and :py:class:`sodym.dimensions.DimensionSet`.
+    """
+
+    def __init__(self, data_reader: DataReader, model_cfg: dict={}):
+        """Define and set up the MFA system and load all required data.
+        Does not compute stocks or flows yet."""
+        for k, v in model_cfg.items():  # Allows other model properties to be set as attributes,
+            # which can be acccessed in other class methods.
+            setattr(self, k, v)
```

Comment:
could we please leave this a dictionary? I.e. simply

`self.model_cfg = model_cfg`. 

I think that's more transparent. If you don't like Dictionaries, we could pass a [named tuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) to the init. And we could shorten the argument AND attribute name to cfg. But I'd prefer no internal quiet conversion.

Reply:
Yes, I also think it's nicer to keep it as a dict (or named tuple, no strong feeling between them). Then problem is just that I don't know what will break doing this. I will do it though, and hopefully I will get all of the attributes!